### PR TITLE
Fix parquet write when the input column is nested type containing timestamp

### PR DESCRIPTION
**Pull request by ttnghia**
This fixes a bug in parquet writer when the input column contains a nested timestamp column but that is not considered for conversion to the correct timestamp format. The bug is described in more detail in https://github.com/NVIDIA/spark-rapids/issues/6302.

Closes https://github.com/NVIDIA/spark-rapids/issues/6302.

### CODE REVIEW

**Comment by ttnghia** [comment]
build

File: sql-plugin/src/main/java/com/nvidia/spark/rapids/ColumnCastUtil.scala [hunk 1]
```diff
@@ -43,16 +43,17 @@ object ColumnCastUtil extends Arm {
...
    * @param cv the view to be updated
+   * @param dt the Spark's data type of the input view (if applicable)
    * @param convert the partial function used to convert the data. If this matches and returns
    *                a updated view this function takes ownership of that view.
    * @return None if there were no changes to the view or the updated view along with anything else
    *         that needs to be closed.
    */
-  def deepTransformView(cv: ColumnView)
-      (convert: PartialFunction[ColumnView, ColumnView]):
+  def deepTransformView(cv: ColumnView, dt: Option[DataType] = None)
```
Comment (revans2) [inline sql-plugin/src/main/java/com/nvidia/spark/rapids/ColumnCastUtil.scala:52]: Don't we need to update the recursive calls to deepTransformView below to include the child DataTypes if it was set? I think nested DecimalType columns are not going to be properly updated.

Comment (ttnghia) [reply sql-plugin/src/main/java/com/nvidia/spark/rapids/ColumnCastUtil.scala:52]: Wait, `DataTypes` can also be nested? I thought not.

Comment (ttnghia) [reply sql-plugin/src/main/java/com/nvidia/spark/rapids/ColumnCastUtil.scala:52]: So if we have a nested type `GpuColumnVector` like `struct<int, double>`, is its `dataType` also a nested struct like `StructType<IntegerType, DoubleType>`?

Comment (revans2) [reply sql-plugin/src/main/java/com/nvidia/spark/rapids/ColumnCastUtil.scala:52]: Mostly.  [StructType](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/types/StructType.scala) has a  `fileds` array that holds a StructField, which holds the name and type of each child. ListType holds the child type it wraps and MapType holds a key type and a value type.

When we create a GpuColumnVector from a cudf.ColumnVector we have asserts to verify that the types match what is expected, but those asserts are off in production. You can assume that they will match, and if they do not you will probably get a class cast exception trying to cast a DataType to the wrong child type.

**Comment by ttnghia** [comment]
build

**Comment by revans2** [comment]
The tests are failing with errors about some of the columns being closed too many times.

**Comment by ttnghia** [comment]
Yes, I just realized that the test is failing because `DType.LIST` also needs to handle `BinaryType`. Testing locally before pushing....

File: sql-plugin/src/main/java/com/nvidia/spark/rapids/ColumnCastUtil.scala [hunk 2]
```diff
@@ -63,12 +64,27 @@ object ColumnCastUtil extends Arm {
           cv.getType.getTypeId match {
             case DType.DTypeEnum.STRUCT =>
               withResource(ArrayBuffer.empty[ColumnView]) { tmpNeedsClosed =>
+                val structFields = if (dt.isDefined) {
```
Comment (revans2) [inline sql-plugin/src/main/java/com/nvidia/spark/rapids/ColumnCastUtil.scala:67]: nit: Something like

```
val structFields = dt match {
  case Some(t: StructType) => t.fields
  case None => Array.empty[StructField]
  case Some(other) => /* this should never be reached */
    throw new IllegalStateException("Invalid input DataType: " +
      s"Expect StructType but got ${t.toString}")
}
```

File: sql-plugin/src/main/java/com/nvidia/spark/rapids/ColumnCastUtil.scala [hunk 3]
```diff
@@ -89,8 +105,25 @@ object ColumnCastUtil extends Arm {
...
               }
             case DType.DTypeEnum.LIST =>
-              withResource(cv.getChildColumnView(0)) { dataView =>
-                val (updatedData, needsClosingData) = deepTransformView(dataView)(convert)
+              withResource(cv.getChildColumnView(0)) { child =>
+                // A ColumnView of LIST type may have data type is ArrayType or MapType in Spark.
+                // If it is a MapType, its child will be a column of type struct<key, value>.
+                // In such cases, we need to generate the corresponding Spark's data type
+                // for the child column as a StructType.
+                val childDt = if (dt.isDefined) {
```
Comment (revans2) [inline sql-plugin/src/main/java/com/nvidia/spark/rapids/ColumnCastUtil.scala:113]: Similar here, you can rewrite this do that the Some is stripped out as a part of the match. It is very minor.

**Comment by ttnghia** [comment]
build

**Comment by ttnghia** [comment]
build

**Comment by ttnghia** [comment]
build

**Comment by ttnghia** [comment]
build

**Review by revans2 — APPROVED** [review]

**Pull request merged**